crypto/bls/signature_batch.go

```diff
@@ -33,12 +45,58 @@ func (s *SignatureBatch) Verify() (bool, error) {
 	return VerifyMultipleSignatures(s.Signatures, s.Messages, s.PublicKeys)
 }
 
+// VerifyVerbosely verifies signatures as a whole at first, if fails, fallback
+// to verify each single signature to identify invalid ones.
+func (s *SignatureBatch) VerifyVerbosely() (bool, error) {
+	valid, err := s.Verify()
+	if err != nil || valid {
+		return valid, err
+	}
+
+	// if signature batch is invalid, we then verify signatures one by one.
+
+	var sb strings.Builder
+
+	for i := 0; i < len(s.Signatures); i++ {
+		sig := s.Signatures[i]
+		msg := s.Messages[i]
+		pubKey := s.PublicKeys[i]
+
+		valid, err := VerifySignature(sig, msg, pubKey)
+		if !valid {
+			var desc string
+			if len(s.Descriptions) != len(s.Signatures) {
+				desc = "non-specific signature"
+			} else {
+				desc = s.Descriptions[i]
+			}
+
+			if err != nil {
+				_, e := fmt.Fprintf(&sb, "signature '%s' is invalid."+
+					" signature: %v, public key: %v, message: %v, error: %v\n", desc, sig, pubKey, msg, err)
+				if e != nil {
```

Comment:
don't ignore errors here, also why are we using `Fprintf` here ? Wrap the error and then return it.

Reply:
Here I used `strings.Builder` for better performance (not a big problem though). `strings.Builder` follows `StringWriter` interface that emits error but actually always returns `nil` as error, so I decided to ignore it. I'll consider to use string concatenation instead.